### PR TITLE
First stab at howto-label

### PR DESCRIPTION
**Pull request by robdodson**
work-in-progress, still need to add tests and tune up the comments. I think the implementation is ready for review though. @surma do you have time to give it a look? Also, what do you think is the best way to pull in another howto-component into my demo?

### CODE REVIEW

File: elements/howto-label/howto-label.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+      const value = this.getAttribute('for');
+      return value === null ? '' : value;
+    }
+
+    set for(value) {
+      this.setAttribute('for', value);
+    }
+
+    attributeChangedCallback(name, oldVal, newVal) {
+      if (name === 'for') {
```
Comment (surma) [inline elements/howto-label/howto-label.js]: Do we wanna do this even when only observing one attribute?

File: elements/howto-label/howto-label.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      return ['for'];
+    }
+
+    constructor() {
+      super();
+      this.attachShadow({mode: 'open'});
+      this.shadowRoot.appendChild(template.content.cloneNode(true));
+      // Listen for the slotchange event to label any wrapped children
+      this.shadowRoot.querySelector('slot')
+        .addEventListener('slotchange', this._forChanged.bind(this));
```
Comment (surma) [inline elements/howto-label/howto-label.js]: It’s a bit weird to call `_forChanged` on `slotchange`. Maybe you can find a better name?

File: elements/howto-label/howto-label.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        target.removeAttribute('aria-labelledby');
+      }
+      target = this._findTarget();
+      target.setAttribute('aria-labelledby', this.id);
+    }
+
+    _onClick(event) {
+      let el = this._currentLabelTarget();
+      // If nothing is labeled, or if a wrapped child was clicked on, return.
+      if (!el || event.target === el) {
```
Comment (surma) [inline elements/howto-label/howto-label.js]: Does this work with inner nested children? Like 

```html
<howto-label>
  <howto-checkbox></howto-checkbox>
  The <em>Label for this <strong>awesome</strong></em> checkbox
</howto-label>
```

I feel like it does because bubbling, but wanted to double-check.

File: elements/howto-label/howto-label.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+     * If there is a for property, return the element with that id.
+     * Else, return the explicitly labeled child.
+     * Else, return the first element child (assume it's implicitly labeled).
+     */
+    _findTarget() {
+      if (this.for) {
+        // external target
+        let scope = this.getRootNode();
+        return scope.getElementById(this.for);
+      } else {
```
Comment (surma) [inline elements/howto-label/howto-label.js]: I’d prefer no `else`

File: elements/howto-label/howto-label.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        let slottedChildren = this.shadowRoot
+          .querySelector('slot').assignedNodes({flatten: true});
```
Comment (surma) [inline elements/howto-label/howto-label.js]: Worth caching the slot?

File: elements/howto-label/howto-label.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        let el;
+        slottedChildren.forEach(child => {
+          // Ignore text nodes
+          if (child.nodeType === Node.TEXT_NODE) {
+            return;
+          }
+          // Implicit internal target, first element child
+          if (!el) {
```
Comment (surma) [inline elements/howto-label/howto-label.js]: This `if` is a bit confusing. Couldn’t you do it like this?

```js
// Get all non-text slotted children
let slottedChildren = 
  this._slot.assignedNodes({flatten: true})
    .filter(child => child.nodeType !== Node.TEXT_NODE);
// Find the first one that defines an explicit external target
let el =
  slottedChildren
    .find(child => child.hasAttribute('howto-label-target'));
// Return that first explicit external target, or the first child if there is none.
return el || slottedChildren[0];
``

**Comment by surma** [comment]
> Also, what do you think is the best way to pull in another howto-component into my demo?

Idk honestly. I’m wondering if we should make `<howto-label>` the exception and include it in every demo.

File: elements/howto-label/howto-label.js (hunk 3 above)
Comment (robdodson) [reply elements/howto-label/howto-label.js]: If you wrap additional element children then you have to put the `howto-label-target` attribute on the checkbox. Otherwise there's no way to know which element u want to be the label target.

File: elements/howto-label/howto-label.js (hunk 6 above)
Comment (robdodson) [reply elements/howto-label/howto-label.js]: yeah but now you're looping through the children twice :P

**Comment by robdodson** [comment]
leaving a note for myself that I need to style `:host` to use the default cursor instead of the text selection one. This is what `<label>` seems to do.

File: elements/howto-label/howto-label.js (hunk 3 above)
Comment (surma) [reply elements/howto-label/howto-label.js]: Understood.

File: elements/howto-checkbox/howto-checkbox.js [hunk 7]
```diff
@@ -30,7 +30,7 @@
     <style>
       :host {
         display: inline-block;
-        background: url('{%PATH%}/images/unchecked-checkbox.svg') no-repeat;
```
Comment (surma) [inline elements/howto-checkbox/howto-checkbox.js]: Is this not working or why did you remove the placeholder?

File: elements/howto-label/howto-label.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      let slottedChildren = this._slot.assignedNodes({flatten: true});
+      let el;
+      slottedChildren.forEach(child => {
+        // Ignore text nodes
+        if (child.nodeType === Node.TEXT_NODE) {
+          return;
+        }
+        // Find the first element child. This is the implicit target
+        // of the label.
+        if (!el) {
```
Comment (surma) [inline elements/howto-label/howto-label.js]: Still not very happy about this bit. I see your point but I think looping twice is a small price to pay for the readability here (and the number of children should be <10 on average, no?).

I won’t block on this, it just seems a bit weird (and harder to understand) to give the first iteration a special code branch.

File: elements/howto-checkbox/howto-checkbox.js (hunk 7 above)
Comment (robdodson) [reply elements/howto-checkbox/howto-checkbox.js]: oh I did it so I could use the element in my demo.html temporarily during testing. Still not sure what would be the best way to pull the checkbox into my demo.

File: elements/howto-label/howto-label.js (hunk 8 above)
Comment (robdodson) [reply elements/howto-label/howto-label.js]: Changing to your previously pitched approach to make the code easier to understand.

**Comment by robdodson** [comment]
@surma PTAL. If you're cool with the e2e changes i can merge and we can take up doing a bigger e2e refactor down the line.
